models/management.cattle.io.feature.js

```diff
@@ -34,6 +34,9 @@ export default class Feature extends HybridModel {
       enabled: state ? this.canDisable : this.canUpdate,
     };
 
+    // Can't disable or enable if the feature flag is locked
+    enableAction.enabled = enableAction.enabled && !this.status.lockedValue;
```

Comment:
Not sure if I read it correctly, but does this mean the action will be enabled if `lockedValue` is `false`?
From what I understand `lockedValue` is a nullable bool, so if it's `null` it means the value is not locked, but if it's `true` or `false` it represents the current locked value - at least judging by this:
https://github.com/rancher/rancher/blob/a8a484badf106ee366497b10346985466120ea9d/pkg/api/norman/customization/feature/feature.go#L42-L44

Reply:
If is a nullable bool, then the code as is, is fine. If it is going to be a string value of "true" or "false", then it will need changing.

Can you point me at the documentation that user's will refer to for how to lock feature flags?

Reply:
@nwmac , it's nullable bool. For example, this is how it looks like for RKE2 feature flag:
```json
"status": {
  "default": true,
  "description": "Enable provisioning of RKE2",
  "dynamic": false,
  "lockedValue": true
}
```

And this is the status for legacy feature flag:
```json
"status": {
  "default": false,
  "description": "Enable legacy features",
  "dynamic": true,
  "lockedValue": null
}
```

Now, I'm still not convinced the code "as is" is fine.
* If we have `lockedValue` set to `null` we're fine as `!this.status.lockedValue` evaluates to `true` and we will be able to change the feature flag value (as expected).
* If we have `lockedValue` set to `true` we're fine as `!this.status.lockedValue` evaluates to `false` and we will not be able to change the feature flag value (as expected).
* If we have `lockedValue` set to `false` we're not fine as `!this.status.lockedValue` evaluates to `true` and we will be able to change the feature flag value. This is not expected as having `lockedValue` set to `false` means that the feature flag is locked to "off", not that the value is not locked.

Please let me know if I'm missing something in my explanation.

As for _the documentation that user's will refer to for how to lock feature flags_ - it's not something users can do. The `lockedValue` is set on backend depending on state of the cluster. For example, in the referenced PR we're setting `lockedValue` to `true` for RKE2 feature flag if there are RKE2 provisioned cluster to prevent the admins from turning off this feature flag. If there are no RKE2 provisioned clusers, `lockedValue` is `null` meaning that the admins can turn it off/on as necessary.

Hope this is helpful.

Reply:
I get it now - thanks for the explanation. I'll need to make a change. Does lockedValue override the default value? Currently we look at `spec.value` for the value and fall back to `status.default` if `spec.value` is not set. Should we be falling back to `status.lockedValue` and then `status.default` ?

Reply:
Yes, this makes sense. In fact, we may even go for `status.lockedValue` having the highest priority (if it's not null) since `status` represents the current state is vs. `spec` being a desired state. But only if it's indeed null since for example Legacy flag I see now has
```
"spec": {
        "value": true
    },
    "status": {
        "default": false,
        "description": "Enable legacy features",
        "dynamic": true,
        "lockedValue": null
    }
```
and we want it to be "ON" in this case and available to be switched off.